ironfish/src/wallet/scanner/walletScanner.ts

```diff
@@ -137,71 +167,43 @@ export class WalletScanner {
       )
     }
 
-    const accounts = await AsyncUtils.filter(this.wallet.listAccounts(), async (account) => {
-      if (!account.scanningEnabled) {
-        return false
-      }
-
-      const accountHead = await account.getHead()
-
-      if (!accountHead) {
-        return blockHeader.sequence === 1
-      } else {
-        return BufferUtils.equalsNullable(accountHead.hash, blockHeader.previousBlockHash)
-      }
-    })
-
-    const shouldDecryptAccounts = accounts.filter((a) =>
-      this.wallet.shouldDecryptForAccount(blockHeader, a),
-    )
-
-    const shouldDecryptAccountIds = new Set(shouldDecryptAccounts.map((a) => a.id))
-
-    const decryptedTransactions = await Promise.all(
-      getTransactionsWithNoteIndex(blockHeader, transactions).map(
-        ({ transaction, initialNoteIndex }) =>
-          this.wallet
-            .decryptNotes(transaction, initialNoteIndex, false, shouldDecryptAccounts)
-            .then((r) => ({
-              result: r,
-              transaction,
-            })),
-      ),
-    )
-
-    // account id -> transaction hash -> Array<DecryptedNote>
-    const decryptedNotesMap: Map<string, BufferMap<Array<DecryptedNote>>> = new Map()
-    for (const { transaction, result } of decryptedTransactions) {
-      for (const [accountId, decryptedNotes] of result) {
-        const accountTxnsMap =
-          decryptedNotesMap.get(accountId) ?? new BufferMap<Array<DecryptedNote>>()
-        accountTxnsMap.set(transaction.hash(), decryptedNotes)
-        decryptedNotesMap.set(accountId, accountTxnsMap)
+    const accounts = new Array<Account>()
+    for (const candidate of this.scanningAccounts) {
+      if (
+        !candidate.scanFrom ||
+        BufferUtils.equalsNullable(candidate.scanFrom.hash, blockHeader.previousBlockHash)
+      ) {
+        candidate.scanFrom = null
+        accounts.push(candidate.account)
       }
     }
 
-    for (const account of accounts) {
-      if (abort?.signal.aborted) {
-        return
-      }
+    return decryptor.decryptNotesFromBlock(
+      blockHeader,
+      transactions,
+      accounts,
+      async (decryptedNotesMap) => {
```

Comment:
Random thought. This is sort of an old school NodeJS approach to use a callback like this now that we have promises. Only one place is really going to use this API but generally the preferred way would probably have this method return Promise<AccountTransactionNoteMap | null> in the case of cancelation.

That being said, I don't want to recommend a change to this because it works great.

Reply:
I'm aware of those patterns, however returning a `Promise` (and, supposedly, awaiting it) in this case can't be done, because it would prevent the next block from being read. The callback is there to make sure that its code gets executed in a different unit without affecting the reading of blocks.